File: legacy/DPB_Comp/matlab/Channel.m
classdef Channel < handle

    properties (Constant)
        speedOfLightInMetersPerSecond = 3e8
    end

    properties
        frequencyInHertz = -1
        transmitAntennaGain = 1
        receiveAntennaGain = 1
    end

    methods (Abstract)
        calculateGain(thisObject, distanceInMeters)
    end

    methods
        function newObject = Channel(frequencyInHertz)
            if frequencyInHertz <= 0
                error('Frequency must be positive');
            end
            newObject.frequencyInHertz = frequencyInHertz;
        end

        function wavelength = getWavelengthInMeters(thisObject)
            c = thisObject.speedOfLightInMetersPerSecond;
            f = thisObject.frequencyInHertz;
            wavelength = c./f;
        end
    end
end

File: legacy/DPB_Comp/python/Channel.py
from abc import ABC, abstractmethod


class Channel(ABC):

    speed_of_light_in_meters_per_second = 3e8

    def __init__(self, frequency_in_hertz):
        self.frequency_in_hertz = frequency_in_hertz
        self.transmit_antenna_gain = 1
        self.receive_antenna_gain = 1

    @property
    def wavelength_in_meters(self):
        c = self.speed_of_light_in_meters_per_second
        f = self.frequency_in_hertz

        lambda_ = c / f

        return lambda_

    @abstractmethod
    def calculate_gain(self, distance_in_meters):
        pass
